tools/Modules/TestFx-Tasks.psm1

```diff
@@ -1,4 +1,160 @@
+
+        [Parameter(ParameterSetName='UserIdParamSet', Mandatory=$true, HelpMessage = "UserId (OrgId) you would like to use")]
+        [ValidateNotNullOrEmpty()]
+        [string]$UserId,
+
+        [Parameter(ParameterSetName='SpnParamSet', Mandatory=$true, HelpMessage='ServicePrincipal/ClientId you would like to use')]   
+        [ValidateNotNullOrEmpty()]
+        [string]$ServicePrincipalId,
+
+        [Parameter(ParameterSetName='SpnParamSet', Mandatory=$true, HelpMessage='ServicePrincipal Secret/ClientId Secret you would like to use')]
+        [ValidateNotNullOrEmpty()]
+        [string]$ServicePrincipalSecret,
+
+        [Parameter(ParameterSetName='CreateSpnParamSet', Mandatory=$true, HelpMessage = "SubscriptionId you would like to use")]
+        [Parameter(ParameterSetName='SpnParamSet', Mandatory=$true, HelpMessage = "SubscriptionId you would like to use")]
+        [Parameter(ParameterSetName='UserIdParamSet', Mandatory=$true, HelpMessage = "SubscriptionId you would like to use")]
+        [ValidateNotNullOrEmpty()]
+        [string]$SubscriptionId,
+
+        [Parameter(ParameterSetName='CreateSpnParamSet', Mandatory=$true, HelpMessage='AADTenant/TenantId you would like to use')]
+        [Parameter(ParameterSetName='SpnParamSet', Mandatory=$true, HelpMessage='AADTenant/TenantId you would like to use')]
+        [ValidateNotNullOrEmpty()]
+        [string]$TenantId,
+
+        [Parameter(ParameterSetName='CreateSpnParamSet', Mandatory=$true, HelpMessage = "SubscriptionId you would like to use")]
+        [Parameter(ParameterSetName='SpnParamSet', Mandatory=$true, HelpMessage = "Record tests, Playback tests, or neither")]
+        [Parameter(ParameterSetName='UserIdParamSet', Mandatory=$true, HelpMessage = "Record tests, Playback tests, or neither")]
+        [ValidateSet("Playback", "Record", "None")]
+        [string]$RecordMode,
+
+        [ValidateSet("Prod", "Dogfood", "Current", "Next", "Custom")]
+        [string]$TargetEnvironment='Prod',
+		
+		[string]$ResourceManagementUri,
+		[string]$GraphUri,
+		[string]$AADAuthUri,
+		[string]$AADTokenAudienceUri,
+		[string]$GraphTokenAudienceUri,		
+		[string]$IbizaPortalUri,
+		[string]$ServiceManagementUri,
+		[string]$RdfePortalUri,
+		[string]$GalleryUri,
+		[string]$DataLakeStoreServiceUri,
+        [string]$DataLakeAnalyticsJobAndCatalogServiceUri,
+        
+        [switch]$Force
+    )
+
+    [hashtable]$credentials = @{}
+    $credentials.SubscriptionId = $SubscriptionId
+    $credentials.HttpRecorderMode = $RecordMode
+    $credentials.Environment = $TargetEnvironment
+
+    if ([string]::IsNullOrEmpty($NewServicePrincipalDisplayName) -eq $false) {
+        $Scope = "/subscriptions/" + $SubscriptionId
+        $SecureStringSecret = ConvertTo-SecureString $NewServicePrincipalSecret -AsPlainText -Force
+        $NewServicePrincipal = New-AzureRMADServicePrincipal -DisplayName $NewServicePrincipalDisplayName -Password $SecureStringSecret
```

Comment:
How do we ensure the user is logged in, and using the correct environment before we create an SPN?